### DISCUSSION TITLE
Microsoft / Github should stop treating open-source software as malware and forcing developers to pay for certificates

### DISCUSSION THREAD

**Discussion by aleksey-hoffman** [community/community#4293] [Other Feature Feedback, Questions, & Ideas]
Currently, all developers of open-source apps are forced to pay $300 - $700 / year for an EV signed certificate, if they want their app to not trigger the Smart Screen telling users their app is dangerous, during the installation process.

The whole point of open-source apps is to be inherently more trusted, since their code can be reviewed by anyone at any moment, yet Windows Smart Screen is teaching users that open-source apps shouldn't be trusted.

Yes, open-source software can theoretically contain hidden malware, but couldn't we at least create a system where reviewed, trusted open-source apps would not trigger the Smart Screen designed to discourage users from installing noname unsigned close-sourced apps?

What about the Microsoft's and Github's statements about their "commitment to empower open-source developers"?
If Microsoft and Github resolved this issue, everyone would benefit - the users, the developers, the Microsoft itself.

I'm the dev of an open-source app called [Sigma File Manager](https://github.com/aleksey-hoffman/sigma-file-manager) and right now I'm, as all other open-source developers, being forced by Microsoft, in particular the Windows Smart Screen prompt, to pay for a certificate so my open-source app wouldn't be considered as harmful as any other untrusted close-sourced app. This is teaching users that open-source apps shouldn't be trusted.

### Method 1:
Microsoft could create a system that would allow open-source developers to apply for a free signed certificate from them.

### Method 2:
When installing an app without a certificate, Windows OS should determine whether this app is open-source or not. And if it is, instead of showing the standard Smart Screen prompt saying the app is dangerous and shouldn't be installed, it should show a more sensible warning:

"This app is open-source. Its code is opened for everyone to review. It hasn't been reviewed by any authoritative company, but it's been reviewed by 578 trusted developers on Github"

With 3 buttons:

"read the source code" | "install" | "do not install"

Even though most people wouldn't be able to read the source code themselves, it would at least teach the users that open-source software is safer than random unsigned closed-source noname apps that haven't been reviewed by anyone. The problem with Windows Smart Screen is that it currently treats both apps exactly the same.

### Method 3:
Microsoft could make all reviewed, popular open-source apps trusted by default. This is a more difficult approach but it can be done.
The certificate could be re-issued on every new release, after a review to make sure no harmful code was added.

Microsoft could integrate this check directly into their Windows OS:

When an app is being installed:

- Make sure the app is open-source by checking if the file exists (has the same hash) on Github or any other trusted open-source platform as an asset of the corresponding project.
- Check if the project has been reviewed and approved by the platform. The community can help here as well, Perhaps, in this case, it would make sense to create a system where trusted users with long history could approve open-source apps. Surely there's a way to distinguish fake apps uploaded a few minutes ago from the real ones with thousands of stars from users with accounts created 1+ year ago, and other identifiers that can help make sure it's a real app (e.g. owner behavior and interactions with other people, PRs, etc).

**Comment by Andre601** [comment]
GitHub is not Microsoft.
Yes they where aquired by them but you can (and should) still see them as an independent company and not as a child of Microsoft that does all their bidding.

This issue is something you should forward to Microsoft directly.
Yes, the license thing is annoying for many, but this has been for years now and there are companies that also offer licenses (which Microsoft to my knowledge supports) for way cheaper.

Next: How should Windows detect and differentiate open source from closed source or even malware software?
This isn't as easy as checking a line or sending a request to the GitHub API to see if the entire code or whatever is on GitHub. A simple comparison of names wouldn't be sufficient enough as malware can adobt such a name and make itself look normal (which is what malware does by default).

Like what you suggest requires a complete, direct connection to the entirety of GitHub and some advanced code/software scanning to see if GitHub even has something similar, which at least for me would not be efficient and only bloat stuff up.

Option 2 is a better approach and would be something similar to what companies like JetBrains offer, which is a one-year Open Source license which is granted under the requirement that it is only used for the development of open source project and without the goal of monetary gain. The license also needs to be renewed every year which is as simple as making the staff check the repository again.

However, it may suffer from a minor downside, which would also be the case with Option 3:
It costs time and human resources.

This is by far the biggest problem to see here. It would take a lot of time and probably hundreds of employees to review each project for being legit to then hand over the license. And that pretty much every day for the entire year.
Not to mention that it may also require employees that know the language(s) (both code but also spoken language) to spot possible backdoors or similar. I mean you can't just let a community supporter or whatever deal with license requests. He needs to know some coding stuff or he might allow a resource with a nasty backdoor to get a free license because it's open source.

----

I don't want to say that your complaint is bad.
All I want to say is, that GitHub is the least good place to request such a change. There might be a chance that they will contact Microsoft to address such an idea, but I still think it's somewhat better to find better places to request this more directly from Microsoft.

**Comment by sylveon** [comment]
Two slight mistakes here:

- Your open source app is not permanently flagged. Once the app is downloaded and ran enough times, it will skip SmartScreen. You can also report a false positive to MS to expedite the process.
- Signatures (unless they are EV) do not ensure SmartScreen is skipped. My open source program tripped SmartScreen as well when I first released a signed update. Future updates also do trigger SmartScreen, but the download/run threshold for the warning to go away is much lower than unsigned because reputation is carried over by the certificate.

Shipping via the Windows Store also does not trigger SmartScreen, because Microsoft reviews those ;)

**Reply by aleksey-hoffman** [reply]
I've noticed that after a few dozens - hundreds of installations the Smart Screen stops appearing. But a 100 people would still have to go through this Smart Screen warning every time the app is updated.

Windows Store is a good solution. but it complicates the development process.

**Reply by sylveon** [reply]
I don't know, I found to be making a MSIX much simpler than building a classical installer

**Reply by aleksey-hoffman** [reply]
I'm developing a cross-platform app, so I'd have to disable certain features (like auto updater) and handle some features differently in the Windows build, if it was distributed via the Windows store

**Reply by sylveon** [reply]
Windows Store is your auto updater in this case.
